File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/AttachmentBar/AttachmentUploadBar.m
//
//  AttachmentUploadBar.m
//  samplechat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "AttachmentUploadBar.h"
#import "UIImage+fixOrientation.h"
#import "UIView+Chat.h"

@implementation AttachmentUploadBar

//MARK: - Life Cycle
- (void)awakeFromNib {
    [super awakeFromNib];
    
    self.cancelButton.hidden = YES;
    self.progressBar.hidden = YES;
    [self.attachmentImageView setRoundViewWithCornerRadius:8.0f];
    self.attachmentImageView.contentMode = UIViewContentModeScaleAspectFill;
    [self setRoundBorderEdgeColorView:0.0f borderWidth:0.5f color:nil borderColor:[UIColor colorWithRed:0.8f green:0.8f blue:0.8f alpha:1.0f]];
}

- (IBAction)cancelButtonPressed:(UIButton *)sender {
    self.cancelButton.hidden = YES;
    if ([self.delegate respondsToSelector:@selector(attachmentBar:didTapCancelButton:)]) {
        [self.delegate attachmentBar:self didTapCancelButton:sender];
    }
}

- (void)updateLoadingProgress:(CGFloat)progress {
    if (self.progressBar.isHidden == YES) {
        self.progressBar.hidden = NO;
    }
    if (progress > 0.0) {
        __weak typeof(self) weakSelf = self;
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf.progressBar setProgressTo:progress];
        });
    }
}

/**
 *  This method is called when the user finishes picking attachment image.
 *
 *  @param image    image that was picked by user
 */
- (void)uploadAttachmentImage:(UIImage *)image pickerControllerSourceType:(UIImagePickerControllerSourceType)sourceType {
    
    self.attachmentImageView.image = image;
    
    __weak typeof(self)weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        __typeof(weakSelf)strongSelf = weakSelf;
        
        UIImage *newImage = image;
        if (sourceType == UIImagePickerControllerSourceTypeCamera) {
            newImage = [newImage fixOrientation];
        }
        
        CGFloat largestSide = image.size.width > image.size.height ? image.size.width : image.size.height;
        CGFloat scaleCoefficient = largestSide / 560.0f;
        CGSize newSize = CGSizeMake(image.size.width / scaleCoefficient, image.size.height / scaleCoefficient);
        
        UIGraphicsBeginImageContext(newSize);
        
        [image drawInRect:(CGRect){0, 0, newSize.width, newSize.height}];
        UIImage *resizedImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        if (resizedImage == nil) {
            return;
        }
        NSData *imageData = UIImagePNGRepresentation(resizedImage);
        // Sending attachment to the dialog.
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [QBRequest TUploadFile:imageData
                          fileName:@"test.png"
                       contentType:@"image/png"
                          isPublic:NO
                      successBlock:^(QBResponse * _Nonnull response, QBCBlob * _Nonnull uploadedBlob) {
                
                QBChatAttachment *attachment = [[QBChatAttachment alloc] init];
                attachment.ID = uploadedBlob.UID;
                attachment.name = uploadedBlob.name;
                attachment.type = @"image";
                
                self.progressBar.hidden = YES;
                self.cancelButton.hidden = NO;
                if ([strongSelf.delegate respondsToSelector:@selector(attachmentBar:didUpLoadAttachment:)]) {
                    [strongSelf.delegate attachmentBar:strongSelf didUpLoadAttachment:attachment];
                }
                
            } statusBlock:^(QBRequest * _Nonnull request, QBRequestStatus * _Nonnull status) {
                
                CGFloat progress = status.percentOfCompletion;
                [strongSelf updateLoadingProgress:progress];
                
            } errorBlock:^(QBResponse * _Nonnull response) {
                if ([strongSelf.delegate respondsToSelector:@selector(attachmentBarFailedUpLoadImage:)]) {
                    [strongSelf.delegate attachmentBarFailedUpLoadImage:strongSelf];
                }
                
            }];
        });
    });
}

@end

File: sample-chat-swift/sample-chat-swift/Components/ChatScreen/Chat/CustomUI/AttachmentBar/AttachmentUploadBar.swift
//
//  AttachmentUploadBar.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/8/20.
//  Copyright © 2020 quickBlox. All rights reserved.
//

import UIKit

protocol AttachmentBarDelegate: class {
    func attachmentBarFailedUpLoadImage(_ attachmentBar: AttachmentUploadBar);
    func attachmentBar(_ attachmentBar: AttachmentUploadBar, didUpLoadAttachment  attachment: QBChatAttachment)
    func attachmentBar(_ attachmentBar: AttachmentUploadBar, didTapCancelButton: UIButton)
}

class AttachmentUploadBar: UIView {

    @IBOutlet weak var cancelButton: UIButton!
    @IBOutlet weak var progressBar: CircularProgressBar!
    @IBOutlet weak var attachmentImageView: UIImageView!
    
    //MARK: - Properties
    weak var delegate: AttachmentBarDelegate?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        cancelButton.isHidden = true
        progressBar.isHidden = true
        attachmentImageView.setRoundedView(cornerRadius: 8.0)
        attachmentImageView.contentMode = .scaleAspectFill
        self.setRoundBorderEdgeColorView(cornerRadius: 0.0, borderWidth: 0.5, borderColor: #colorLiteral(red: 0.8039215803, green: 0.8039215803, blue: 0.8039215803, alpha: 1))
        
    }
    
    //MARK: - Actions
    /**
     *  This method is called when the user finishes picking attachment image.
     *
     *  @param image    image that was picked by user
     */
    func uploadAttachmentImage(_ asset: PhotoAsset, url: URL? = nil, attachmentType: AttachmentType) {
        
        attachmentImageView.image = asset.image;
        
        DispatchQueue.global().async { [weak self] () -> Void in
            
            guard let self = self else { return }
            if let url = url {
                let name = FileManager.default.displayName(atPath: url.path)
                if attachmentType == .Video {
                    DispatchQueue.main.async(execute: {
                        QBRequest.uploadFile(with: url, fileName: name, contentType: "video/mp4", isPublic: true,
                                             successBlock: { [weak self] (response: QBResponse, uploadedBlob: QBCBlob) -> Void in
                                                guard let self = self else {
                                                    return
                                                }
                                                let attachment = QBChatAttachment()
                                                attachment.id = uploadedBlob.uid
                                                attachment.name = uploadedBlob.name
                                                attachment.type = "video"
                                                attachment["size"] = "\(uploadedBlob.size)"
                                                
                                                self.progressBar.isHidden = true
                                                self.cancelButton.isHidden = false
                                                self.delegate?.attachmentBar(self, didUpLoadAttachment: attachment)
                                                
                            }, statusBlock: { [weak self] (request : QBRequest?, status : QBRequestStatus?) -> Void in
                                if let status = status {
                                    DispatchQueue.main.async {
                                        let progress = CGFloat(status.percentOfCompletion)
                                        self?.updateLoadingProgress(progress)
                                    }
                                }
                        }) { [weak self] (response : QBResponse) -> Void in
                            guard let self = self else {
                                return
                            }
                            self.delegate?.attachmentBarFailedUpLoadImage(self)
                        }
                    })
                    
                } else if attachmentType == .File {
                    DispatchQueue.main.async(execute: {
                        QBRequest.uploadFile(with: url, fileName: name, contentType: "application/pdf", isPublic: true,
                                             successBlock: { [weak self] (response: QBResponse, uploadedBlob: QBCBlob) -> Void in
                                                guard let self = self else {
                                                    return
                                                }
                                                let attachment = QBChatAttachment()
                                                attachment.id = uploadedBlob.uid
                                                attachment.name = uploadedBlob.name
                                                attachment.type = "file"
                                                attachment["size"] = "\(uploadedBlob.size)"
                                                
                                                self.progressBar.isHidden = true
                                                self.cancelButton.isHidden = false
                                                self.delegate?.attachmentBar(self, didUpLoadAttachment: attachment)
                                                
                            }, statusBlock: { [weak self] (request : QBRequest?, status : QBRequestStatus?) -> Void in
                                if let status = status {
                                    DispatchQueue.main.async {
                                        let progress = CGFloat(Float(status.percentOfCompletion))
                                        self?.updateLoadingProgress(progress)
                                    }
                                }
                        }) { [weak self] (response : QBResponse) -> Void in
                            guard let self = self else {
                                return
                            }
                            self.delegate?.attachmentBarFailedUpLoadImage(self)
                        }
                    })
                }
                
            } else {
                
                var name = "Test.png"
                if let fileName = asset.phAsset.value(forKey: "filename") as? String{
                    name = fileName
                }
                
                var newImage = asset.image
                if attachmentType == .Camera {
                    newImage = newImage.fixOrientation()
                }
                
                let largestSide = newImage.size.width > newImage.size.height ? newImage.size.width : newImage.size.height
                let scaleCoeficient = largestSide/560.0
                let newSize = CGSize(width: newImage.size.width/scaleCoeficient, height: newImage.size.height/scaleCoeficient)
                
                // create smaller image
                UIGraphicsBeginImageContext(newSize)
                
                newImage.draw(in: CGRect(x: 0, y: 0, width: newSize.width, height: newSize.height))
                let resizedImage = UIGraphicsGetImageFromCurrentImageContext()
                
                UIGraphicsEndImageContext()
                
                // Sending attachment.
                DispatchQueue.main.async(execute: {
                    guard let imageData = resizedImage?.pngData() else {
                        return
                    }
                    QBRequest.tUploadFile(imageData, fileName: name,
                                          contentType: "image/png", isPublic: true,
                                          successBlock: { [weak self] (response: QBResponse, uploadedBlob: QBCBlob) -> Void in
                                            guard let self = self else {
                                                return
                                            }
                                            let attachment = QBChatAttachment()
                                            attachment.id = uploadedBlob.uid
                                            attachment.name = uploadedBlob.name
                                            attachment.type = "image"
                                            attachment["size"] = "\(uploadedBlob.size)"
                                            
                                            self.progressBar.isHidden = true
                                            self.cancelButton.isHidden = false
                                            self.delegate?.attachmentBar(self, didUpLoadAttachment: attachment)
                                            
                        }, statusBlock: { [weak self] (request : QBRequest?, status : QBRequestStatus?) -> Void in
                            if let status = status {
                                DispatchQueue.main.async {
                                    let progress = CGFloat(Float(status.percentOfCompletion))
                                    self?.updateLoadingProgress(progress)
                                }
                            }
                    }) { [weak self] (response : QBResponse) -> Void in
                        guard let self = self else {
                            return
                        }
                        self.delegate?.attachmentBarFailedUpLoadImage(self)
                    }
                })
            }
        }
    }
    
    func updateLoadingProgress(_ progress: CGFloat) {
        if progressBar.isHidden == true {
            progressBar.isHidden = false
        }
        progressBar.setProgress(to: progress)
    }
    
    @IBAction func cancelButtonPressed(_ sender: UIButton) {
        cancelButton.isHidden = true
        delegate?.attachmentBar(self, didTapCancelButton: sender)
    }
    
}
